mono/mini/method-to-ir.c

```diff
@@ -8940,6 +8940,30 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				}
 			}
 
+			// Optimize
+			// 
+			//   box 
+			//   call object::GetType()
+			//
+			guint32 gettype_token;
+			if ((ip = il_read_call(next_ip, end, &gettype_token)) && ip_in_bb (cfg, cfg->cbb, ip)) {
+				MonoMethod* gettype_method = mini_get_method (cfg, method, gettype_token, NULL, generic_context);
+				if (!strcmp (gettype_method->name, "GetType") && gettype_method->klass == mono_defaults.object_class) {
+					mono_class_init_internal(klass);
+					MonoType *klass_type = m_class_get_byval_arg (klass);
+					if (!mono_class_is_open_constructed_type (klass_type) && !cfg->compile_aot) { // FIXME: implement for AOT
+						// Emit `typeof(T)`
```

Comment:
There is a
								EMIT_NEW_TYPE_FROM_HANDLE_CONST (cfg, ins, m_class_get_image (tclass), m_class_get_type_token (tclass), generic_context);
runtime constant which will work for aot as well.
It should be used like this:
if (mono_class_get_checked (m_class_get_image (tclass), m_class_get_type_token (tclass), error) == tclass)
								EMIT_NEW_TYPE_FROM_HANDLE_CONST (cfg, ins, m_class_get_image (tclass), m_class_get_type_token (tclass), generic_context);
To filter out classes which have no type token like arrays.